test/sslapitest.c

```diff
@@ -1865,35 +1871,238 @@ static int ocsp_server_cb(SSL *s, void *arg)
         return SSL_TLSEXT_ERR_ALERT_FATAL;
     }
 
-    if (!TEST_ptr(copy = OPENSSL_memdup(orespder, sizeof(orespder))))
+    dummy_ocsp_resp = OCSP_response_create(OCSP_RESPONSE_STATUS_SUCCESSFUL, NULL);
+
+    if (!TEST_ptr(dummy_ocsp_resp))
         return SSL_TLSEXT_ERR_ALERT_FATAL;
 
-    if (!TEST_true(SSL_set_tlsext_status_ocsp_resp(s, copy,
-                                                   sizeof(orespder)))) {
-        OPENSSL_free(copy);
+    resplen = i2d_OCSP_RESPONSE(dummy_ocsp_resp, &ocspresp);
+
+    OCSP_RESPONSE_free(dummy_ocsp_resp);
+
+    if (!TEST_true(SSL_set_tlsext_status_ocsp_resp(s, ocspresp, resplen))) {
+        OPENSSL_free(ocspresp);
         return SSL_TLSEXT_ERR_ALERT_FATAL;
     }
+
+    OPENSSL_free(ocspresp);
+
     ocsp_server_called = 1;
     return SSL_TLSEXT_ERR_OK;
 }
 
-static int ocsp_client_cb(SSL *s, void *arg)
+static int ocsp_client_cb_single(SSL *s, void *arg)
 {
     int *argi = (int *)arg;
-    const unsigned char *respderin;
-    size_t len;
+    const unsigned char *resp, *p;
+    OCSP_RESPONSE *rsp;
+    int len, resp_status;
 
     if (*argi != 1 && *argi != 2)
         return 0;
 
-    len = SSL_get_tlsext_status_ocsp_resp(s, &respderin);
-    if (!TEST_mem_eq(orespder, len, respderin, len))
+    len = SSL_get_tlsext_status_ocsp_resp(s, &resp);
+
+    p = resp;
+    rsp = d2i_OCSP_RESPONSE(NULL, &p, len);
+
+    resp_status = OCSP_response_status(rsp);
+
+    OCSP_RESPONSE_free(rsp);
+
+    if (resp_status != OCSP_RESPONSE_STATUS_SUCCESSFUL)
         return 0;
 
     ocsp_client_called = 1;
     return 1;
 }
 
+static int ocsp_server_cb_multi(SSL *s, void *arg)
+{
+    int *argi = (int *)arg;
+    const SSL_CONNECTION *sc = SSL_CONNECTION_FROM_CONST_SSL(s);
+    ASN1_TIME *thisupd = NULL, *nextupd = NULL;
+    X509 *ssl_cert = NULL;
+    X509 *issuer = NULL;
+    int i, num = 0;
+    STACK_OF(X509) *server_certs = NULL;
+    STACK_OF(OCSP_RESPONSE) *sk_resp = NULL;
+    OCSP_CERTID *cert_id = NULL;
+    OCSP_BASICRESP *bs = NULL;
+    char *signer_key_file = test_mk_file_path(certsdir, "subinterCA.key");
+    char *signer_cert_file = test_mk_file_path(certsdir, "subinterCA.pem");
+    BIO *bio_signer_key = NULL;
+    EVP_PKEY *signer_key = NULL;
+    X509 *signer_cert = NULL;
+    const EVP_MD *md = EVP_MD_fetch(libctx, "SHA-256", NULL);
+
+    if (*argi != 1 && *argi != 2)
+        return 0;
+
+    SSL_get0_chain_certs(s, &server_certs);
+
+    if (server_certs != NULL && sc != NULL) {
+        /* certificate chain is available */
+        num = sk_X509_num(server_certs);
+    } else if (sc != NULL) {
+        /*
+         * certificate chain is not available,
+         * set num to 1 for server certificate
+         */
+        num = 1;
+    }
+
+    if (*argi == 2)
+        num = 1;
```

Comment:
```suggestion
    if (num != *argi)
        return 0;
```
for more intuitive use of the arg (see also below).